csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +263,335 @@ int64_t getMaxPersistentBatch(
+      max_threads_per_block,
+      ", but got ",
+      params.bdimx);
+  params.bdimy = std::min(
+      scheduler_utils::safeDiv(min_threads_per_block, params.bdimx),
+      max_multi_reduction_factor);
+  params.padded_bdimx = params.bdimx % device_warp_size == 0
+      ? params.bdimx
+      : params.bdimx + (device_warp_size - params.bdimx % device_warp_size);
+  params.is_pad_bdimx = params.bdimx > 16 &&
+      params.padded_bdimx * params.bdimy <= max_threads_per_block;
+
+  // calculate register per thread and achieved occupancy
+  int64_t threads_per_block = params.is_pad_bdimx
+      ? params.padded_bdimx * params.bdimy
+      : params.bdimx * params.bdimy;
+  int64_t persistent_buffer_size =
+      buffer_bytes_per_batch * persistent_batch_size;
+  auto reg_occ = getMaxRegisterCountPerThreadAndOccupancy(
+      persistent_buffer_size,
+      threads_per_block,
+      target_warps_per_sm,
+      free_registers);
+  params.register_per_thread = reg_occ.first;
+  params.occupancy = reg_occ.second;
+  params.free_registers = params.register_per_thread -
+      persistent_buffer_size / scheduler_utils::bytes_per_register;
+  // (4) Calculate other quantities reflecting the quality of the heuristic.
+  // when [reduction_count_after_vectorize] is not divisible by
+  // [persistent_val], the last batch is not be fully utilized, the wasted
+  // threads in the last batch is quantified as [n_persistent_tails].
+  params.n_persistent_tails =
+      ceilDiv(reduction_count_after_vectorize, persistent_batch_size) *
+          persistent_batch_size -
+      reduction_count_after_vectorize;
+  return params;
+}
+
+// Return true if ha is better than hb
+
+// This sorting function ensures the selected heuristic meeting target
+// occupancy, promotes even workload distribution, enhances register
+// optimization, and prefers higher occupancy.
+// TODO: It leads to 10% regression for softmax around 2K to 6K and 16K.
+// See https://github.com/NVIDIA/Fuser/issues/1876
+bool compareTwoHeuristics(
+    const HeuristicParams& ha,
+    const HeuristicParams& hb,
+    const int64_t target_register_overhead,
+    const int64_t target_warps_per_sm) {
+  auto compare = [](int64_t a, int64_t b) -> int {
+    return a > b ? 1 : (a < b ? -1 : 0);
+  };
+  int score = 0;
+
+  // prefer occupancy larger than target
+  score = compare(
+      ha.occupancy >= target_warps_per_sm, hb.occupancy >= target_warps_per_sm);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // prefer reduction count after vectorization is divisible by persistent
+  // batch size
+  score = compare(ha.n_persistent_tails == 0, hb.n_persistent_tails == 0);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // Large register overhead leaves more registers for the compiler to
+  // optimize generated SASS code and avoids register spills. But don't want
+  // to use a very large block size which increaes inter-thread communication.
```

Comment:
I don't disagree that larger block sizes would have larger overheads, but why should this particular condition be used? Why not, for example, just having something like `compare(ha.padded_bdimx < hb.padded_bdimx)`? 

If there's any backing reasoning, please leave it as a comment. If not, then please mention that it is experimentally found to work well.

Reply:
Agree on this one: https://github.com/NVIDIA/Fuser/pull/1849#discussion_r1523952768